main.go

```diff
@@ -0,0 +1,145 @@
+package main
+
+import (
+	"flag"
+	"log"
+	"net/http"
+	"os/user"
+	"strconv"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"github.com/tkanos/gonfig"
+)
+
+const (
+	DOMAIN_KEY = "domain"
+	CONFIG     = "catchall.json"
+)
+
+func serveHttp(
+	wg *sync.WaitGroup, mode string, queue Queue,
+	backendRead BackendReader, port int) {
+	defer wg.Done()
+	router := mux.NewRouter()
+	// Write mode (PUTs)
+	if mode == "RW" || mode == "W" {
+		router.Handle(
+			"/events/{"+DOMAIN_KEY+"}/delivered",
+			HandleRequest(false, queue)).Methods("PUT")
+		router.Handle(
+			"/events/{"+DOMAIN_KEY+"}/bounced",
+			HandleRequest(true, queue)).Methods("PUT")
+	}
+	// Read mode (GETs)
+	if mode == "RW" || mode == "R" {
+		router.Handle(
+			"/domains/{"+DOMAIN_KEY+"}", HandleQuery(backendRead)).Methods("GET")
+	}
+	router.PathPrefix("/").HandlerFunc(CatchAllHandler)
+
+	// NOTE(roaet): Possible performance increase with valyala/fasthttp
+	log.Fatal(http.ListenAndServe(":"+strconv.Itoa(port), router))
+}
+
+func doWork(wg *sync.WaitGroup, queue Queue, backendWrite BackendWriter) {
+	defer wg.Done()
+	for {
+		log.Println("Worker waiting for work")
+		entries, err := queue.pull(1)
+		if err != nil {
+			// TODO(roaet): probably shouldn't exit
+			log.Fatal("Error pulling from the queue")
+		}
+		if len(entries) == 0 {
+			continue
+		}
+		for _, entry := range entries {
+			backendWrite.insert(entry)
+		}
+	}
+}
+
+type Configuration struct {
+	Port               int
+	Worker, API, Local bool
+	APIMode            string
+	CassandraHosts     []string
+	RabbitURI          string
+}
+
+func initializeConfig(filepath string, config *Configuration) error {
+	usr, err := user.Current()
+	if err != nil {
+		log.Fatalf("Could not load user information: %s\n", err)
+	}
+	homedir := usr.HomeDir
+	err = gonfig.GetConf(homedir+"/"+CONFIG, config)
+	if err != nil {
+		log.Fatalf("Could not load config: %s: %s\n", CONFIG, err)
+	}
+	return nil
+}
+
+func main() {
```

Comment:
Handling signals would be nice.

Reply:
Signal handling would allow the process to die gracefully as suggested by aerwin3 (who I can't tag).